### PR TITLE
csi: add CSIDriverOptions section in cephCluster CR

### PR DESCRIPTION
**Pull request by Rakshith-R**
This commit adds new CSIDriverOptions section in
cephCluster CR. This section contains settings
for read affinity and kernel+fuse Mount options
These settings will be injected directly into
rook-ceph-csi-config cm to be applicable per
ceph cluster.

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

### CODE REVIEW

File: deploy/examples/cluster.yaml
Comment (Madhu-1) [inline deploy/examples/cluster.yaml:285]: how abou the below CR template

```code
csi:
   general:
        x: x
   cephfs:
        x:x
   rbd:
      x:x
   nfs:
     x:x
```

The above might helps us to keep the config specific and easily extendable?

**Comment by travisn** [comment]
> We recently added COSI to Rook, and in that, we decided to have the COSI driver options be its own CR. I would suggest that we take this approach for CSI as well. This will keep us from over-complicating the CephCluster CR more than it already is. It will keep consistency between CSI and COSI drivers.
> 
> Also, I think this makes more sense when considering Rook as a whole: Rook runs one CSI driver per k8s cluster regardless of how many CephClusters are present, so I think it makes most logical sense to have a separate CR for it. We wouldn't want to have the CephCluster that configured the CSI driver to be deleted and then to have an additional CephCluster then have its CSI settings changed or removed as a result.

The CSI settings being considered are cluster-specific. Any configuration of these settings would not affect other clusters upon creation or deletion. During the cluster reconcile, Rook creates/updates a configmap specific for a given cluster with settings related to that cluster. For example, the mon endpoints are included in the configmap. The proposal with these settings is that they would also be specific to the cluster, and would be added to the configmap only for that cluster. Note the configmap is in the operator namespace, but has a section for each cluster. 

Since the configmap is already updated during the cephcluster reconcile, and these proposed settings would be cluster-specific, we could choose either to have a CRD for the csi settings, or add them to the cephcluster CR. If we define a new CRD, the cluster controller could reconcile it, rather than implementing a new controller. I could see the new CRD being more clear from the user perspective, even if under the covers the two approaches are functionally the same.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:230]: Is there a struct defined in the CSI repo that we could use in the CRD type? If Rook uses the CSI type directly, we don't need to translate between the Rook and CSI types. When we read from the CR, it can be immediately serialized to the configmap exactly as the csi driver expects it, and no translation code would be needed. Perhaps that won't be possible, but would like to see if it's an option.

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (Rakshith-R) [reply pkg/apis/ceph.rook.io/v1/types.go:230]: > Is there a struct defined in the CSI repo that we could use in the CRD type? If Rook uses the CSI type directly, we don't need to translate between the Rook and CSI types. When we read from the CR, it can be immediately serialized to the configmap exactly as the csi driver expects it, and no translation code would be needed. Perhaps that won't be possible, but would like to see if it's an option.

Currently the struct is not exported. https://github.com/ceph/ceph-csi/blob/593d9c3b605adc742940a012736184662315983d/internal/util/csiconfig.go#L40.

This needs to be done at cephcsi first and changes will need to be done at a lot of places in rook repo.
I'll open a separate follow up pr to import structs from cephcsi repo and use it in rook.
Even with the struch, some translation would definitely be required.
That change will be internal only, we don't need to block this pr.

**Comment by Rakshith-R** [comment]
> > We recently added COSI to Rook, and in that, we decided to have the COSI driver options be its own CR. I would suggest that we take this approach for CSI as well. This will keep us from over-complicating the CephCluster CR more than it already is. It will keep consistency between CSI and COSI drivers.
> > Also, I think this makes more sense when considering Rook as a whole: Rook runs one CSI driver per k8s cluster regardless of how many CephClusters are present, so I think it makes most logical sense to have a separate CR for it. We wouldn't want to have the CephCluster that configured the CSI driver to be deleted and then to have an additional CephCluster then have its CSI settings changed or removed as a result.
> 
> The CSI settings being considered are cluster-specific. Any configuration of these settings would not affect other clusters upon creation or deletion. During the cluster reconcile, Rook creates/updates a configmap specific for a given cluster with settings related to that cluster. For example, the mon endpoints are included in the configmap. The proposal with these settings is that they would also be specific to the cluster, and would be added to the configmap only for that cluster. Note the configmap is in the operator namespace, but has a section for each cluster.
> 
> Since the configmap is already updated during the cephcluster reconcile, and these proposed settings would be cluster-specific, we could choose either to have a CRD for the csi settings, or add them to the cephcluster CR. If we define a new CRD, the cluster controller could reconcile it, rather than implementing a new controller. I could see the new CRD being more clear from the user perspective, even if under the covers the two approaches are functionally the same.

Apart from the above points: CSI settings being cluster-specific and configmap already being filled during cephcluster reconcile. 

Having a separate CR would lead to unnecessary duplication of code and complicate the logic further. 
We just inject the settings into the configmap and there's no other functionality for this separate CR. 
It would complicate current code that injects mon ips and radosnamespace/subvolumegroup too.

I'd prefer we keep the csi driver settings within cephcluster CR (a single source of truth).

File: deploy/examples/cluster.yaml
Comment (Rakshith-R) [reply deploy/examples/cluster.yaml:285]: @Madhu-1 Can you please check it now ? 
I've segregated settings into common and cephfs for now.

**Comment by BlaineEXE** [comment]
I chatted with Travis and now understand that there are CephCluster-specific settings for CSI. I misunderstood initially, so thanks for your patience @Rakshith-R .

I also think that keeping the configs with the CephCluster makes sense in that light. I don't think we gain anything by using a separate CRD.

File: deploy/examples/cluster.yaml
Comment (travisn) [inline deploy/examples/cluster.yaml:296]: I don't know why anyone would ever need to override this since if they care about topology, they should already be using the same topology labels that rook already uses. So I'd suggest we not even include the crushLocationLabels in the cluster.yaml example, we can just have it in the schema and the docs.

File: deploy/examples/cluster.yaml
Comment (travisn) [inline deploy/examples/cluster.yaml:304]: We should set `ms_mode=secure` automatically if encryption is enabled. They should only set this if they need to override the defaults, right?

File: deploy/examples/cluster.yaml
Comment (travisn) [inline deploy/examples/cluster.yaml:307]: Is fuse supported now, or still just experimental with ceph csi?

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [reply pkg/apis/ceph.rook.io/v1/types.go:232]: I guess we need different structs anyway since there are differences where Rook will generate the mons, and we obviously don't want those specified in the cluster CR under the csi settings.

File: pkg/operator/ceph/csi/cluster_config.go
Comment (travisn) [inline pkg/operator/ceph/csi/cluster_config.go:43]: Seems like this struct should be deleted and the CSI struct referenced instead. But as you said, this can be a separate PR.

File: pkg/operator/ceph/file/subvolumegroup/controller.go
Comment (travisn) [inline pkg/operator/ceph/file/subvolumegroup/controller.go:276]: Where are the Rook topology labels added by default? Seems like they are not added by Rook. Does the CSI driver have the Rook labels as defaults or Rook needs to set them?

**Review by travisn — CHANGES_REQUESTED** [review]

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [reply pkg/apis/ceph.rook.io/v1/types.go:232]: I think it would be challenging to use the whole CSI config struct directly in the CephCluster. I imagine that many of the fields don't make sense for users to modify since Rook will fill them in dynamically based on the Ceph cluster config and status.

If there are any specific data types that CSI uses that have special validation, it could be useful for CSI to provide those as an API that Rook can use to pre-validate configs before they reach CSI. I've noticed that a lot of users struggle with CSI debugging when they have mis-configured something because they only look at Rook operator logs and events rather than CSI logs and events, and I think that would be helpful.

And also, that can certainly be something that gets small improvements over time.

File: deploy/examples/cluster.yaml
Comment (BlaineEXE) [reply deploy/examples/cluster.yaml:287]: I strongly plus-one this. Any configs that are shared by all CSI drivers should not exist on CephClusters. With 'common' settings, different CephClusters could compete to be the source of truth for the shared CSI settings. Competition is easily solved by keeping common settings separate, in a singleton. Currently, the CSI common config is the rook-ceph-operator-config configmap, but it could be a separate CRD in the future, like COSI has done.

File: deploy/examples/cluster.yaml
Comment (BlaineEXE) [reply deploy/examples/cluster.yaml:296]: I wonder if we should also just strike this from the schema entirely and make it so users cannot modify this. It seems to me that it might be best if Rook handles this via its own mechanisms internally. With this in the schema, it seems more likely that a user might set this in such a way that drastically conflicts with what Rook has done with the Ceph cluster. But perhaps there is a use for this that I don't quite understand.

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (BlaineEXE) [inline pkg/apis/ceph.rook.io/v1/types.go:228]: As it is here, this spec item is not nullable because it is not a pointer to a struct.

Should this be a pointer to struct (`*CSIDriverOptionsSpec`), or should `// +nullable` be removed?

File: pkg/operator/ceph/csi/cluster_config.go
Comment (BlaineEXE) [reply pkg/operator/ceph/csi/cluster_config.go:43]: Plus one

File: pkg/operator/ceph/file/subvolumegroup/controller.go
Comment (BlaineEXE) [reply pkg/operator/ceph/file/subvolumegroup/controller.go:276]: I'm also curious how the topology levels affect read affinity. Should Rook be configured to always send Ceph-CSI its list of topology labels every time? 

I have a sense that doing so might mean that Rook can change its topology labels without Ceph-CSI needing to be updated to get the new label changes. I don't foresee that being necessary as a result of Ceph, but I do see that it could be reasonable for a Rook user to want to configure Rook to look at custom topology keys rather than only use the built-in ones. If that were the case, Rook might also want to send an updated list of topology labels automatically. 

I think this also captures more of my thoughts around the topology labels from my comment here: https://github.com/rook/rook/pull/13200/files#r1395017616

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (Rakshith-R) [reply pkg/apis/ceph.rook.io/v1/types.go:228]: > As it is here, this spec item is not nullable because it is not a pointer to a struct.
> 
> Should this be a pointer to struct (`*CSIDriverOptionsSpec`), or should `// +nullable` be removed?

:+1: 
`nullable` was added by mistake, removed it.

File: pkg/operator/ceph/file/subvolumegroup/controller.go
Comment (Rakshith-R) [reply pkg/operator/ceph/file/subvolumegroup/controller.go:276]: Modified code to supply the default set of crushlocation labels now similar to how the global cmdline args for read affinity is handled. (:sweat_smile: got confused earlier and had missed it)

There is a case for having crushlocationlabels option exposed: 
User maybe using Rook to connect to an external stretch cluster which maybe spread across zone Z1 and Z2, Z1 being the closer one. 
Custom labels will help user tell the ceph client to get reads from the nearest zone Z1 from the external cluster
and not clash with those of existing default topology labels.

File: deploy/examples/cluster.yaml
Comment (Rakshith-R) [reply deploy/examples/cluster.yaml:307]: > Is there any special behavior users or Rook need to handle here? Does setting `fuseMountOptions` automatically select the FUSE driver? It might be good to note that for users regardless of the answer.
> 
> Either, "setting fuseMountOptions automatically selects the FUSE driver," or "fuseMountOptions only applies when the FUSE driver is enabled"
> 
> I also presume that since I don't see kernel/fuse mount options here, that means that kernel/fuse enable/disable is a common setting for all ceph clusters?

These settings only pass the mountoptions globally for the cephfs driver in a single ceph cluster. 
Enabling disabling/ auto selection depends on the storage class(cluster specific).

Comment (Rakshith-R) [reply deploy/examples/cluster.yaml:307]: > Is fuse supported now, or still just experimental with ceph csi?

There are still downsides but it is in better shape.

File: deploy/examples/cluster.yaml
Comment (Rakshith-R) [reply deploy/examples/cluster.yaml:304]: > We should set `ms_mode=secure` automatically if encryption is enabled. They should only set this if they need to 
override the defaults, right?

That line is from https://github.com/rook/rook/blob/f86fca5e7ad620d54798f12a10adf88baa0663fb/deploy/examples/operator.yaml#L487
That is still something user needs to do. Doing so automatically may cause existing clients to get blocked.
https://github.com/rook/rook/pull/11625. 

cc @Madhu-1

File: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md
Comment (travisn) [inline Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:255]: How about also showing the yaml snippet for readability in case they prefer to edit their manifests? 
```yaml
  csi:
    readAffinity:
      enabled: true
```

File: deploy/examples/cluster.yaml
Comment (travisn) [inline deploy/examples/cluster.yaml:288]: Links aren't rendered in the yaml, how about this? We want to link to the rendered docs as well, instead of the .md.
```suggestion
# For more details see this doc:
# https://rook.io/docs/rook/latest/Storage-Configuration/Ceph-CSI/ceph-csi-drivers/#enable-read-affinity-for-rbd-volumes
```

File: deploy/examples/cluster.yaml
Comment (travisn) [reply deploy/examples/cluster.yaml:295]: We couldn't do it automatically in the past because of the issue with ceph clusters that might have different settings. Now with the cluster settings being independent, seems like we can set it automatically when encryption is enabled. This setting would just allow them to override the defaults.

File: deploy/examples/cluster.yaml
Comment (travisn) [reply deploy/examples/cluster.yaml:298]: curious to discuss more another time on this topic.

**Review by travisn — CHANGES_REQUESTED** [review]
The CI is failing

**Comment by Rakshith-R** [comment]
>We couldn't do it automatically in the past because of the issue with ceph clusters that might have different settings. Now with the cluster settings being independent, seems like we can set it automatically when encryption is enabled. This setting would just allow them to override the defaults. 

I am not sure how it might affect certain cases like upgrades or toggling network encryption status. 
I've opened a new issue #13227 to discuss this. We can take up this change in a follow-up pr.

File: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md
Comment (travisn) [inline Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:267]: How about we remove this operator config from the docs? Even if we still have it as a fallback/default setting, we want people to start using the new setting, and no need for anyone to use it, right?

File: deploy/examples/cluster.yaml
Comment (travisn) [inline deploy/examples/cluster.yaml:287]: Let's tell them basically about the feature so they don't always have to go to the doc.
```suggestion
      # Enable read affinity to enable clients to optimize reads from an OSD in the same topology.
```

File: deploy/examples/cluster.yaml
Comment (travisn) [inline deploy/examples/cluster.yaml:290]: Remind me why this isn't enabled by default? If there is some concern to enable it, we should tell the user here.

**Review by travisn — CHANGES_REQUESTED** [review]

File: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md
Comment (Rakshith-R) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md]: > How about we remove this operator config from the docs? Even if we still have it as a fallback/default setting, we want people to start using the new setting, and no need for anyone to use it, right?

Makes sense, Its removed now.

File: deploy/examples/cluster.yaml
Comment (Rakshith-R) [reply deploy/examples/cluster.yaml:290]: > Remind me why this isn't enabled by default? If there is some concern to enable it, we should tell the user here.

There is increased OSD memory usage since multiple OSDs in different zones will be serving the same piece of data for clients in that zone instead of the data coming from the same primary OSD.

File: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md
Comment (parth-gr) [inline Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: Mentioning this isn't require anymore?

File: deploy/examples/cluster.yaml
Comment (travisn) [reply deploy/examples/cluster.yaml:290]: Ok, how about adding a comment to the setting here? Even if we don't have specific numbers, something like this seems helpful: 
```
# Enabling the read affinity will cause the OSDs to consume some extra memory.
```

File: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md
Comment (travisn) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: good catch, seems necessary

**Review by travisn — CHANGES_REQUESTED** [review]

File: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md
Comment (Rakshith-R) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: Actually no, That step was needed for external ceph cluster for explicitly specifying the nearest crush location.

We already adopt whatever topology labels are used for internal ceph cluster for read affinity.
We've already documented that advanced configuration here https://github.com/rook/rook/blob/e54f112f04aefdbc6016fae819729c6d8f139818/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-topology

There's no special steps needed for internal ceph cluster.

Comment (parth-gr) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: Should we document this at external cluster document then?

Comment (Rakshith-R) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: :+1:  that part will require some elaborate explanation.
I've opened a new doc issue to track this.

https://github.com/rook/rook/issues/13243

Comment (parth-gr) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: Thanks!

File: pkg/operator/ceph/pool/radosnamespace/controller.go
Comment (parth-gr) [inline pkg/operator/ceph/pool/radosnamespace/controller.go:267]: is cephfs mounts update needed with rbd radosnamspace?

File: pkg/operator/ceph/csi/cluster_config.go
Comment (parth-gr) [inline pkg/operator/ceph/csi/cluster_config.go:340]: Does ceph CSI  already have the changes to handle the new config?

File: pkg/operator/ceph/pool/radosnamespace/controller.go
Comment (Rakshith-R) [reply pkg/operator/ceph/pool/radosnamespace/controller.go:267]: > is cephfs mounts update needed with rbd radosnamspace?

well the same cluster id can be used for cephfs storageclass too, so its better to have it.

File: pkg/operator/ceph/csi/cluster_config.go
Comment (Rakshith-R) [reply pkg/operator/ceph/csi/cluster_config.go:340]: > Does ceph CSI already have the changes to handle the new config?

nope, it will come with release 3.10 but this is a backward compatible non-breaking change so it should be fine.

File: pkg/operator/ceph/pool/radosnamespace/controller.go
Comment (parth-gr) [reply pkg/operator/ceph/pool/radosnamespace/controller.go:267]: No but looks little miss-leading

File: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md
Comment (travisn) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: To be clear, the admin owns adding the topology labels to their nodes, so this statement does apply to internal clusters. This is about adding the labels to the nodes, which is different from specifying the labels in the cluster CR which we do automatically. After this is resolved, @BlaineEXE feel free to dismiss my review to merge while i'm out.

Comment (Rakshith-R) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: > To be clear, the admin owns adding the topology labels to their nodes, so this statement does apply to internal clusters. This is about adding the labels to the nodes, which is different from specifying the labels in the cluster CR which we do automatically. After this is resolved, @BlaineEXE feel free to dismiss my review to merge while i'm out.

Both topology labels and labels that need to be added to nodes are the same.

https://github.com/rook/rook/blob/e54f112f04aefdbc6016fae819729c6d8f139818/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-topology

> The topology of the cluster is important in production environments where you want your data spread across failure domains. The topology can be controlled by adding labels to the nodes. When the labels are found on a node at first OSD deployment, Rook will add them to the desired level in the [CRUSH map](https://docs.ceph.com/en/latest/rados/operations/crush-map/).


This needs to be done as part of cluster deployment stage.
So for internal cluster it already exists. Therefore, we don't need to mention it in read affinity section.

File: pkg/operator/ceph/pool/radosnamespace/controller.go
Comment (Rakshith-R) [reply pkg/operator/ceph/pool/radosnamespace/controller.go:267]: > No but looks little miss-leading

I dont think so.
This is completely internal so its fine.

Comment (parth-gr) [reply pkg/operator/ceph/pool/radosnamespace/controller.go:267]: We are upgrading this with `ReconcileCephFilesystemSubVolumeGroup` so wanted to understand why it is needed with radosnamespace, is this so that we don't lose track of it in the config map? @Rakshith-R

Comment (Rakshith-R) [reply pkg/operator/ceph/pool/radosnamespace/controller.go:267]: > We are upgrading this with `ReconcileCephFilesystemSubVolumeGroup` so wanted to understand why it is needed with radosnamespace, is this so that we don't lose track of it in the config map? @Rakshith-R

These settings(read affinity  + mount options) must be considered default from Cephcluster CR and added when ever you create a new entry,

**Review by parth-gr — APPROVED** [review]

File: Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md
Comment (BlaineEXE) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: I think it is still reasonable to add a note that users are expected to have already set up topology labels. It **is** a requirement for this feature, and it is reasonable to note requirements specifically so that we don't hold an assumption that users may be unaware of. Additionally, users can miss things on accident, and a reminder is not harmful. I would suggest changing this to mention that it is a requirement, and link to the section of the deployment doc that provides instructions.

Comment (Rakshith-R) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: > I think it is still reasonable to add a note that users are expected to have already set up topology labels. It is a requirement for this feature, and it is reasonable to note requirements specifically so that we don't hold an assumption that users may be unaware of. Additionally, users can miss things on accident, and a reminder is not harmful. I would suggest changing this to mention that it is a requirement, and link to the section of the deployment doc that provides instructions.

The line existed as part of steps for both internal and external clusters.

For internal ceph cluster, it is not a requirement.

- The [OSD-topology-labels](https://github.com/rook/rook/blob/e54f112f04aefdbc6016fae819729c6d8f139818/Documentation/CRDs/Cluster/ceph-cluster-crd.md#osd-topology) consists of both custom rook/ceph topology labels and kubernetes topoogy labels. 
- These gets passed onto crush map when OSD is first created.
- Even if the user does not specify them manually, the host and k8s topology which are present by default are still applied.

- We really don't need to bother the user if he hasn't applied such elaborate rook specific rack/row/datacenter labels before hand. If he has, those same labels will anyway be used for read affinity and no extra action is required.
- Changing topology and crush maps at a later point after deployment is too complex and probably not required.


I've opened a new doc issue to track doc changes for external cluster.: https://github.com/rook/rook/issues/13243. 

I've modified docs to add a line linking to the full set of topology labels so user can check which topology labels are being referred to in this section.

Comment (BlaineEXE) [reply Documentation/Storage-Configuration/Ceph-CSI/ceph-csi-drivers.md:261]: The topology labels are **not** present by default in all k8s clusters. I still stand by my statement. It is still a requirement that the user set up topology labels before this feature, and users might need reminded of that. Not all users set up topology labels on their K8s clusters. Additionally, the link you mentioned does not document Ceph-only labels. It also includes the inbuilt k8s labels as well, and it does not ask that users fill in more info than they need.

The things that can happen with this documented are:
- users check out the OSD topology labels, see they already set some, and then do nothing
- OR, users check out the OSD topology labels, realize they haven't set them, and then do so, catching a critical mistake they made

The things that can happen with this **removed** from documentation are:
- users check out the OSD topology labels, see they already set some, and then do nothing (same as above)
- users forgot to set topology labels for their clusters, and then they deploy with only the host label, making their topology affinity based on node rather than on zone or whatever else they intend
  - And, because of what you mentioned, "Changing topology and crush maps at a later point after deployment is too complex," it is difficult for users to fix their mistake after having made it

The possible user experiences are worse when this doc is removed. Please leave it in place.

**Review by BlaineEXE — APPROVED** [review]

File: pkg/apis/ceph.rook.io/v1/types.go
Comment (travisn) [inline pkg/apis/ceph.rook.io/v1/types.go:255]:
```suggestion
	// CrushLocationLabels defines which node labels to use
```

**Review by travisn — CHANGES_REQUESTED** [review]
one more typo

**Pull request merged**
